be/src/exec/pipeline/aggregate/aggregate_streaming_sink_operator.cpp

```diff
@@ -89,63 +89,190 @@ Status AggregateStreamingSinkOperator::_push_chunk_by_force_preaggregation(Chunk
+        size_t remain_size = real_capacity - _aggregator->hash_map_variant().size();
+        bool ht_needs_expansion = remain_size < chunk_size;
+        _auto_context.init_preagg_count++;
+        if (!ht_needs_expansion ||
+            _aggregator->should_expand_preagg_hash_tables(_aggregator->num_input_rows(), chunk_size, allocated_bytes,
+                                                          _aggregator->hash_map_variant().size())) {
+            // hash table is not full or allow to expand the hash table according reduction rate
             SCOPED_TIMER(_aggregator->agg_compute_timer());
-            RETURN_IF_ERROR(_aggregator->compute_batch_agg_states(chunk.get(), chunk_size));
+            TRY_CATCH_BAD_ALLOC(_aggregator->build_hash_map(chunk_size));
+            if (_aggregator->is_none_group_by_exprs()) {
+                RETURN_IF_ERROR(_aggregator->compute_single_agg_state(chunk.get(), chunk_size));
+            } else {
+                RETURN_IF_ERROR(_aggregator->compute_batch_agg_states(chunk.get(), chunk_size));
+            }
+
+            _mem_tracker->set(_aggregator->hash_map_variant().reserved_memory_usage(_aggregator->mem_pool()));
+            TRY_CATCH_BAD_ALLOC(_aggregator->try_convert_to_two_level_map());
+
+            COUNTER_SET(_aggregator->hash_table_size(), (int64_t)_aggregator->hash_map_variant().size());
+            break;
         } else {
-            // middle cases, first aggregate locally and output by stream
-            {
-                SCOPED_TIMER(_aggregator->agg_compute_timer());
-                RETURN_IF_ERROR(_aggregator->compute_batch_agg_states_with_selection(chunk.get(), chunk_size));
+            _auto_state = AggrAutoState::ADJUST;
+            _auto_context.adjust_count = 0;
+            LOG(INFO) << "auto agg: " << _auto_context.get_auto_state_string(AggrAutoState::INIT_PREAGG) << " "
+                      << _auto_context.init_preagg_count << " -> " << _auto_context.get_auto_state_string(_auto_state);
+        }
+    }
+    case AggrAutoState::ADJUST: {
+        _auto_context.adjust_count++;
+        {
+            SCOPED_TIMER(_aggregator->agg_compute_timer());
+            TRY_CATCH_BAD_ALLOC(_aggregator->build_hash_map_with_selection(chunk_size));
+        }
+
+        size_t agg_count = SIMD::count_zero(_aggregator->streaming_selection());
+        if (_auto_context.adjust_count < continuous_limit && _auto_context.low_reduction(agg_count, chunk_size)) {
+            RETURN_IF_ERROR(_push_chunk_by_force_streaming(chunk));
+            _auto_context.pass_through_count++;
+            _auto_context.preagg_count = 0;
+            _auto_context.selective_preagg_count = 0;
+            if (_auto_context.pass_through_count == AggrAutoContext::AdjustLimit) {
+                _auto_state = AggrAutoState::PASS_THROUGH;
+                LOG(INFO) << "auto agg: continuous " << AggrAutoContext::AdjustLimit << " low reduction "
```

Comment:
`INFO` maybe too noisy?

Reply:
I will change it to VLOG_ROW after testing.